Port Delete By Query to Reindex infrastructure

tlrx opened the pull request:
This commit ports the Delete By Query implementation on the same infrastructure as the Reindex and Update By Query APIs.

The Delete By Query shares most of its code with the Update By Query feature. Some code has been moved around so that there's no duplication. Reindex does not use it for now but the changes will help that. There's surely some improvements to do but I'd be happy to get some feedback from a review.. @nik9000 Can you please have a look? That'd be awesome.

Closes #16883

nik9000 commented on docs/reference/docs/delete-by-query.asciidoc:
```diff
@@ -0,0 +1,308 @@
+
+experimental[The delete-by-query API is new and should still be considered experimental.  The API may change in ways that are not backwards compatible]
+
+The simplest usage of `_delete_by_query` just performs a deletion on every
```
I think maybe you shouldn't be allowed to `_delete_by_query` with specifying a query? `_update_by_query` is less dangerous I think....

tlrx replied:
I agree, this is a doc issue. The code checks for the presence of a query and do not allow `_delete_by_query` usage without a query.

nik9000 replied:
Cool! I'd just replace the whole sentence with something like "`_delete_by_query` performs a delete on every document in an index that matches a query. For example:" Or something like that.

nik9000 commented on docs/reference/docs/delete-by-query.asciidoc:
```diff
@@ -0,0 +1,308 @@
+and when the delete request is processed. When the versions match the document
+is deleted.
+
+All delete and query failures cause the `_delete_by_query` to abort and are
```
I think this is a part of update-by-query and reindex and delete-by-query's docs that'll have to be changed. Rejected executions are retried for bulk, and, with #18331, will be for searches.

tlrx replied:
Agreed - I changed that... Let me know if it's better.

nik9000 commented on docs/reference/docs/delete-by-query.asciidoc:
```diff
@@ -0,0 +1,308 @@
+returned by the failing bulk request are returned in the `failures` element so
+it's possible for there to be quite a few.
+
+If you want to simply count version conflicts not cause the `_delete_by_query`
```
This sentence is a bit funky - Maybe "If you'd like to count version conflicts rather than cause them to abort the `_delete_by_query` then set..." ? Something like that

nik9000 commented on docs/reference/docs/delete-by-query.asciidoc:
```diff
@@ -0,0 +1,308 @@
+// CONSOLE
+// TEST[setup:twitter]
+
+By default `_delete_by_query` uses scroll batches of 100. You can change the
```
1000 now, I think.

nik9000 commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkIndexByScrollAction.java:
```diff
@@ -40,216 +41,359 @@
     /**
-     * Copies the metadata from a hit to the index request.
+     * Used to accept or ignore a search hit. Ignored search hits will be excluded
+     * from the bulk request.
```
Might want to mention that this is also where we fail on invalid search hits. I expect delete-by-query doesn't need to fail like this though.

nik9000 commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkIndexByScrollAction.java:
```diff
@@ -40,216 +41,359 @@
+     * Copy the routing from a search hit to the action request.
+     */
+    protected void copyRouting(ActionRequest<?> request, String routing) {
+        if (request instanceof IndexRequest) {
```
I'm not a fan of having to do all the instanceof checks but I'm not sure what the best way to avoid them is.

nik9000 replied:
Honestly maybe wrapping the requests in another object - there is one interface and two implementations and they handle either arm of the instanceof check?

tlrx replied:
> I'm not a fan of having to do all the instanceof checks but I'm not sure what the best way to avoid them is.

Yes... This where my sentence in the description of the pull request _"There's surely some improvements to do but I'd be happy to get some feedback"_  makes all its sense :)

> Honestly maybe wrapping the requests in another object - there is one interface and two implementations and they handle either arm of the instanceof check?

There's a `DocumentRequest` class around, maybe I can use it...

nik9000 replied:
If it fits! Good luck!

nik9000 commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkIndexByScrollAction.java:
```diff
@@ -40,216 +41,359 @@
-        extends AbstractAsyncBulkByScrollAction<Request, BulkIndexByScrollResponse> {
+public abstract class AbstractAsyncBulkIndexByScrollAction<Request extends AbstractBulkByScrollRequest<Request>>
+        extends AbstractAsyncBulkByScrollAction<Request>
+        implements BiFunction<ActionRequest<?>, SearchHit, ActionRequest<?>> {
```
I might prefer passing a reference to the function rather rather than implementing it here.

nik9000 commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractBaseReindexRestHandler.java:
```diff
@@ -88,41 +63,95 @@ protected AbstractBaseReindexRestHandler(Settings settings, Client client,
+    /**
+     * Sets common options of {@link AbstractBulkByScrollRequest} requests.
+     */
+    protected Request setCommonOptions(RestRequest restRequest, Request request) {
```
Cool.

nik9000 commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/DeleteByQueryRequest.java:
```diff
@@ -0,0 +1,79 @@
+ * <ul>
+ *     <li>it's <tt>non-atomic</tt>, a delete-by-query may fail at any time while some documents matching the query have already been
+ *     deleted</li>
+ *     <li>it's <tt>try-once</tt>, a delete-by-query may fail at any time and will not retry it's execution. All retry logic is left to
```
This isn't really true. Things like conflicts=ignore and retry on bulk failure and, soon, retry on search failure, all count as retries to some degree.

tlrx replied:
I agree - this is a copy/paste from DBQ plugin.

nik9000 commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/DeleteByQueryRequest.java:
```diff
@@ -0,0 +1,79 @@
+    @Override
+    public ActionRequestValidationException validate() {
+        ActionRequestValidationException e = super.validate();
+        if (getSearchRequest().indices() == null || getSearchRequest().indices().length == 0) {
```
These fell like the same validations as are in update-by-query's request. Maybe a common superclass is in order?

nik9000 commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/RestDeleteByQueryAction.java:
```diff
@@ -0,0 +1,76 @@
+         /*
+         * Passing the search request through DeleteByQueryRequest first allows
+         * it to set its own defaults which differ from SearchRequest's
+         * defaults. Then the parse can override them.
```
s/parse/parseInternalRequest/

nik9000 commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/TransportDeleteByQueryAction.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    /**
+     * Simple implementation of delete-by-query using scrolling and bulk.
```
I'm not sure if the "simple" comment really applies to anything in reindex any more....

nik9000 commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/TransportDeleteByQueryAction.java:
```diff
@@ -0,0 +1,99 @@
+
+        @Override
+        protected boolean accept(SearchHit doc) {
+            return true;
```
May want to comment that you this explicitly disables check for _source.

nik9000 replied:
Maybe delete-by-query can disable requesting the source by default? The same we we enable requesting the version?

tlrx replied:
Good catch!

nik9000 commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/TransportDeleteByQueryAction.java:
```diff
@@ -0,0 +1,99 @@
+
+        @Override
+        protected ActionRequest<?> copyMetadata(ActionRequest<?> request, SearchHit doc) {
+            copyParent(request, fieldValue(doc, ParentFieldMapper.NAME));
```
Maybe also good to comment on how the superclass is much more update oriented and you override to skip all of that.

nik9000 commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/TransportReindexAction.java:
```diff
@@ -57,8 +59,9 @@
     public TransportReindexAction(Settings settings, ThreadPool threadPool, ActionFilters actionFilters,
-            IndexNameExpressionResolver indexNameExpressionResolver, ClusterService clusterService, ScriptService scriptService,
-            AutoCreateIndex autoCreateIndex, Client client, TransportService transportService) {
+                                  IndexNameExpressionResolver indexNameExpressionResolver, ClusterService clusterService,
```
The arguments don't change, right?

tlrx replied:
Right.. revert this one

nik9000 commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/TransportReindexAction.java:
```diff
@@ -168,102 +183,124 @@ protected IndexRequest buildIndexRequest(SearchHit doc) {
             }
             if (routingSpec.startsWith("=")) {
-                index.routing(mainRequest.getDestination().routing().substring(1));
+                super.copyRouting(request, mainRequest.getDestination().routing().substring(1));
```
Nice.

nik9000 commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/TransportReindexAction.java:
```diff
@@ -168,102 +183,124 @@ protected IndexRequest buildIndexRequest(SearchHit doc) {
-                break;
-            default:
-                throw new IllegalArgumentException("Unsupported routing command");
+                case "keep":
```
Awe, but I like my properly formatted switch statements!

tlrx replied:
I reverted this one too. It looks like we're not consistent within the code base regarding `switch` statements

nik9000 replied:
Nope! There are heathens who like it with indented case statements and they are in the majority too! But we don't have rules for it.

nik9000 commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/TransportUpdateByQueryAction.java:
```diff
@@ -78,14 +82,33 @@ protected void doExecute(UpdateByQueryRequest request, ActionListener<BulkIndexB
-        protected IndexRequest buildIndexRequest(SearchHit doc) {
+        public ActionRequest<?> apply(ActionRequest<?> request, SearchHit doc) {
+            ActionRequest<?> result = super.apply(request, doc);
+            if (scriptApplier == null) {
```
This feels like something the superclass should handle. Reindex and update-by-query can pass the applier into the async action and delete-by-query can pass null? Then the superclass doesn't have to implement BiFunction?

tlrx replied:
I agree... I changed that, let me know what you think.

nik9000 commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/TransportUpdateByQueryAction.java:
```diff
@@ -97,44 +120,52 @@ protected IndexRequest buildIndexRequest(SearchHit doc) {
-        @Override
-        protected void scriptChangedTTL(IndexRequest index, Object to) {
-            throw new IllegalArgumentException("Modifying [" + TTLFieldMapper.NAME + "] not allowed");
+        class UpdateByQueryScriptApplier extends ScriptApplier {
```
I like the idea of the script applier!

nik9000 commented on modules/reindex/src/test/java/org/elasticsearch/index/reindex/DeleteByQueryCancelTests.java:
```diff
@@ -0,0 +1,182 @@
+ * places - that is the responsibility of {@link AsyncBulkByScrollActionTests} which have more precise control to simulate failures but do
+ * not exercise important portion of the stack like transport and task management.
+ */
+public class DeleteByQueryCancelTests extends ReindexTestCase {
```
At some point soon I'd like to merge all of these cancel tests into a single test. I hadn't done that at first because ReindexTestCase couldn't build an UpdateByQueryRequestBuilder but that isn't the case any more.

nik9000 commented on modules/reindex/src/test/java/org/elasticsearch/index/reindex/DeleteByQueryCancelTests.java:
```diff
@@ -0,0 +1,182 @@
+
+    @Override
+    protected int numberOfShards() {
+        // Only 1 shard
```
might want to explain why you can only have one shard.

nik9000 commented on modules/reindex/src/test/java/org/elasticsearch/index/reindex/DeleteByQueryCancelTests.java:
```diff
@@ -0,0 +1,182 @@
+        private final CountDown blockAfter = new CountDown(MAX_DELETIONS);
+
+        @Override
+        public Engine.Delete preDelete(Engine.Delete delete) {
```
I like this a lot more than I like the way I do update-by-query and reindex cancellation. Do you think as a followup you can port those to this test?

tlrx replied:
> Do you think as a follow up you can port those to this test?

I think so. I can give it a try once this PR is merged.

nik9000 commented on modules/reindex/src/test/java/org/elasticsearch/index/reindex/DeleteByQueryConcurrentTests.java:
```diff
@@ -0,0 +1,143 @@
+                index("test", "test", String.valueOf(i * 10 + j), "field", j);
+            }
+        }
+        refresh();
```
I'd probably build these with `indexRandom` because it'll usually bulk which is usually faster.

nik9000 commented on modules/reindex/src/test/java/org/elasticsearch/index/reindex/DeleteByQueryConcurrentTests.java:
```diff
@@ -0,0 +1,143 @@
+        assertHitCount(client().prepareSearch("test").setSize(0).get(), docs * threads.length);
+
+        final CountDownLatch start = new CountDownLatch(1);
+        final AtomicReference<Throwable> exceptionHolder = new AtomicReference<>();
```
So it turns out that RandomizedRunner sets an uncaught exception handler which should make this not required. I think.

tlrx replied:
I didn't know about it, thanks!

nik9000 replied:
I didn't know it until a few days ago!

nik9000 commented on modules/reindex/src/test/java/org/elasticsearch/index/reindex/DeleteByQueryConcurrentTests.java:
```diff
@@ -0,0 +1,143 @@
+        final Thread[] threads =  new Thread[scaledRandomIntBetween(2, 5)];
+        final long docs = randomIntBetween(1, 50);
+        for (int i = 0; i < docs; i++) {
+            for (int j = 0; j < threads.length; j++) {
```
Maybe you should make `j` `t` instead.

nik9000 commented on modules/reindex/src/test/java/org/elasticsearch/index/reindex/DeleteByQueryConcurrentTests.java:
```diff
@@ -0,0 +1,143 @@
+            throw assertionError;
+        }
+
+        for (int i = 0; i < threads.length; i++) {
```
And here too, `i` maybe should be `t` instead.

nik9000 commented on modules/reindex/src/test/java/org/elasticsearch/index/reindex/DeleteByQueryConcurrentTests.java:
```diff
@@ -0,0 +1,143 @@
+    }
+
+    public void testConcurrentDeleteByQueriesOnSameDocs() throws Throwable {
+        assertAcked(prepareCreate("test").setSettings(Settings.builder().put("index.refresh_interval", -1)));
```
I've learned that any test that relies on refresh interval is going to fail. Some item something is going to force a refresh on us when we least expect it. Is this test ok with that?

tlrx replied:
It looks like this is not necessary anymore... I removed it.

nik9000 commented on modules/reindex/src/test/java/org/elasticsearch/index/reindex/DeleteByQueryConcurrentTests.java:
```diff
@@ -0,0 +1,143 @@
+        final AtomicLong deleted = new AtomicLong(0);
+
+        for (int i = 0; i < threads.length; i++) {
+            assertHitCount(client().prepareSearch("test").setSize(0).setQuery(query).get(), docs);
```
You probably don't need to do this per thread.

nik9000 commented on modules/reindex/src/test/java/org/elasticsearch/index/reindex/DeleteByQueryConcurrentTests.java:
```diff
@@ -0,0 +1,143 @@
+                    start.await();
+
+                    BulkIndexByScrollResponse response = deleteByQuery().source("_all").filter(query).refresh(true).get();
+                    deleted.addAndGet(response.getDeleted());
```
You might want to comment that you expect some of these requests to fail with a version conflict but that the total will be the total number of docs. Right?

tlrx replied:
True. I added a comment about this.

nik9000 commented:
I like it! I think you should change the way the metadata is copied somewhat - something with less instanceofs would make me much more happy.

I think there is probably a cleaner way to do the scripts so that the async actions don't have to implement BiFunction and apply.

I wonder if you can get `op_type=delete` to work for `update_by_query` and `reindex` now that these are there? It feels like that isn't much more than what you've already done. Maybe another PR after this one?

I think it is worth adding this to the breaking changes list somehow too. The plugin is gone and the output is different, right?

tlrx commented:
Thanks for this great review @nik9000 !

I updated the code according to your comments. Please let me know what you think.

> I like it! I think you should change the way the metadata is copied somewhat - something with less instanceofs would make me much more happy.

I agree. I followed your suggestion and wrapped action requests and exposed an interface to manipulate them because `DocumentRequest` does not fit.

> I think there is probably a cleaner way to do the scripts so that the async actions don't have to implement BiFunction and apply.

Passing a method reference (or Supplier) forces all classes to be `static` as well as `copy*` methods... So I added a `getScriptApplier()` method.


> I wonder if you can get op_type=delete to work for update_by_query and reindex now that these are there? It feels like that isn't much more than what you've already done. Maybe another PR after this one?

That's exactly the idea :) but I think it's better if this is done in a separate PR.



> I think it is worth adding this to the breaking changes list somehow too. The plugin is gone and the output is different, right?

I plan to do it when I'll remove the plugin, if that's OK for you.

nik9000 commented:
>> I like it! I think you should change the way the metadata is copied somewhat - something with less instanceofs would make me much more happy.
> I agree. I followed your suggestion and wrapped action requests and exposed an interface to manipulate them because DocumentRequest does not fit.

Cool! I'll have a look!

>> I wonder if you can get op_type=delete to work for update_by_query and reindex now that these are there? It feels like that isn't much more than what you've already done. Maybe another PR after this one?
> That's exactly the idea :) but I think it's better if this is done in a separate PR.

Awesome.

> I think it is worth adding this to the breaking changes list somehow too. The plugin is gone and the output is different, right?
> I plan to do it when I'll remove the plugin, if that's OK for you.

Great.

nik9000 commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkIndexByScrollAction.java:
```diff
@@ -60,13 +60,29 @@
+     * Return the {@link ScriptApplier} to apply to all {@link ActionRequest}, null if
+     * no {@link ScriptApplier} must be applied.
+     */
+    protected ScriptApplier getScriptApplier() {
```
Rename to `buildScriptApplier` and return a noop implementation by default? I usually prefer that to null checks for things that behave like this.

nik9000 commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/RestDeleteByQueryAction.java:
```diff
@@ -63,14 +63,18 @@ protected DeleteByQueryRequest buildRequest(RestRequest request) throws IOExcept
         parseInternalRequest(internal, request, consumers);
+
+        // Delete-By-Query does not require the source
+        internal.getSearchRequest().source().fetchSource(false);
```
I'd do this in the DeleteByQueryRequest instead so it applies in the java API?

nik9000 commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/TransportDeleteByQueryAction.java:
```diff
@@ -64,18 +68,20 @@ protected void doExecute(DeleteByQueryRequest request, ActionListener<BulkIndexB
         @Override
         protected boolean accept(SearchHit doc) {
+            // Delete-by-query does not require the
+            // source to delete a document
```
"and the default implementation checks for it"

nik9000 commented on modules/reindex/src/test/java/org/elasticsearch/index/reindex/DeleteByQueryConcurrentTests.java:
```diff
@@ -19,41 +19,38 @@
-
         final Thread[] threads =  new Thread[scaledRandomIntBetween(2, 5)];
         final long docs = randomIntBetween(1, 50);
+        int t;
```
I'd probably still declare `t` in the loop just because it is the normal thing to do.

nik9000 commented:
I reviewed again and it looks pretty good! I think maybe rebase to resolve the conflicts I just created by merging #18331 and I'll have another look in my morning with fresh eyes?

tlrx commented:
@nik9000 Thanks again! I rebased and updated the code according to your comments.

nik9000 commented on docs/reference/docs/delete-by-query.asciidoc:
```diff
@@ -0,0 +1,311 @@
+
+[source,js]
+--------------------------------------------------
+DELETE /twitter/_delete_by_query
```
I kind of wonder if `_delete_by_query` should be a `POST` instead of a `DELETE`. You aren't really deleting the thing at the URL so maybe `POST` is more right because you are causing the `_delete_by_query` action to be taken? I dunno.

tlrx replied:
I think the first implementation was `DELETE /_query`. I personally like the `DELETE /_delete_by_query` but let's ask @clintongormley about this :)

clintongormley replied:
I think we should move to `POST`, not least of all because some clients (eg JS) don't allow bodies with a `DELETE` request.  Plus I agree with @nik9000's reasoning here

tlrx replied:
Thanks, let's do that then.

nik9000 commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkIndexByScrollAction.java:
```diff
@@ -40,108 +41,180 @@
+     * from copying search hit metadata (parent, routing, etc) to potentially
+     * transforming the {@link RequestWrapper} completely.
+     */
+    protected RequestWrapper<?> apply(RequestWrapper<?> request, SearchHit doc) {
```
I think maybe this doesn't have to be protected any more? Maybe it doesn't even have to be a method any more and you can just call the scriptApplier directly?

tlrx replied:
Yes, let's remove this completly

nik9000 commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkIndexByScrollAction.java:
```diff
@@ -40,108 +41,180 @@
-    protected void copyRouting(IndexRequest index, SearchHit doc) {
-        index.routing(fieldValue(doc, RoutingFieldMapper.NAME));
+    protected void copyType(RequestWrapper<?> request, String type) {
+        request.setType(type);
```
I wonder if these methods are important any more now that you have the RequestWrapper? Like maybe just call the RequestWrapper method directly?

tlrx replied:
Sure, I should have done that.

nik9000 commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkIndexByScrollAction.java:
```diff
@@ -40,108 +41,180 @@
+     * (if supported by this one)
+     */
+    protected void copyTimestamp(RequestWrapper<?> request, Long timestamp) {
+        copyTimestamp(request, timestamp != null ? timestamp.toString() : null);
```
This method still has some logic but maybe it should just be a final method on the RequestWrapper and it should delegate to an abstract method on RequestWrapper? That'd mean RequestWrapper has to be an abstract class but that is ok.

nik9000 replied:
I think.

nik9000 replied:
Or maybe none of that has to be true because _ttl only supported on the wrapper that wraps Index requests.

tlrx replied:
I removed the `copy*` methods and let the wrapper to handle this correctly.

nik9000 commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkIndexByScrollAction.java:
```diff
@@ -40,108 +41,180 @@
-    private final CompiledScript script;
+    protected final ScriptService scriptService;
+    protected final ClusterState clusterState;
+    private final ScriptApplier scriptApplier;
```
If you made this a BiFunction then you could get away with the noop implementation being just `(a, b) -> a` and save a ton of code! `buildScriptApplier` would have to return a BiFunction as well, but extensions to this class could still build an extension of ScriptApplier just like they do now.

tlrx replied:
Yeah, I hesitated a lot about this. Happy to have your opinion! Let's make `buildScriptApplier`returns the `BiFunction` and that will be more simpler.

nik9000 commented on modules/reindex/src/main/java/org/elasticsearch/index/reindex/RestRethrottleAction.java:
```diff
@@ -39,6 +39,7 @@ public RestRethrottleAction(Settings settings, RestController controller, Client
         super(settings, client);
         this.action = action;
         controller.registerHandler(POST, "/_update_by_query/{taskId}/_rethrottle", this);
+        controller.registerHandler(POST, "/_delete_by_query/{taskId}/_rethrottle", this);
```
I think you also should add this to `reindex.rethrottle.json` in the rest spec.

tlrx replied:
Good catch

nik9000 commented:
I'm getting a test failure with this:
```
==> Test Info: seed=D21734D0466B8A8E; jvm=1; suite=1
Suite: org.elasticsearch.smoketest.SmokeTestReindexWithGroovyIT
  2> REPRODUCE WITH: gradle :qa:smoke-test-reindex-with-groovy:integTest -Dtests.seed=D21734D0466B8A8E -Dtests.class=org.elasticsearch.smoketest.SmokeTestReindexWithGroovyIT -Dtests.method="test {yaml=update_by_query/10_script/Setting bogus ctx is an error}" -Des.logger.level=WARN -Dtests.security.manager=true -Dtests.jvms=12 -Dtests.locale=es-PE -Dtests.timezone=Asia/Dushanbe
FAILURE 0.21s | SmokeTestReindexWithGroovyIT.test {yaml=update_by_query/10_script/Setting bogus ctx is an error} <<< FAILURES!
   > Throwable #1: java.lang.AssertionError: the error message was expected to match the provided regex but didn't
   > Expected: Invalid fields added to ctx \[junk\]
   >      but: was "{root_cause=[{type=illegal_argument_exception, reason=Invalid fields added to context [junk]}], type=illegal_argument_exception, reason=Invalid fields added to context [junk]}"
   >    at __randomizedtesting.SeedInfo.seed([D21734D0466B8A8E:5A430B0AE897E776]:0)
   >    at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
   >    at org.elasticsearch.test.rest.section.DoSection.execute(DoSection.java:117)
   >    at org.elasticsearch.test.rest.ESRestTestCase.test(ESRestTestCase.java:399)
   >    at java.lang.Thread.run(Thread.java:745)
  2> NOTE: leaving temporary files on disk at: /home/manybubbles/Workspaces/Elasticsearch/elasticsearch/qa/smoke-test-reindex-with-groovy/build/testrun/integTest/J0/temp/org.elasticsearch.smoketest.SmokeTestReindexWithGroovyIT_D21734D0466B8A8E-001
  2> NOTE: test params are: codec=Asserting(Lucene60): {}, docValues:{}, maxPointsInLeafNode=1342, maxMBSortInHeap=5.027602980016196, sim=ClassicSimilarity, locale=es-PE, timezone=Asia/Dushanbe
  2> NOTE: Linux 3.16.0-4-amd64 amd64/Oracle Corporation 1.8.0_72-internal (64-bit)/cpus=12,threads=1,free=444417848,total=514850816
  2> NOTE: All tests run in this JVM: [SmokeTestReindexWithGroovyIT]
Completed [1/1] in 8.49s, 21 tests, 1 failure <<< FAILURES!
```


When I want to test "just" the reindex module I run `gradle modules:reindex:check qa:smoke-test-reindex-with-groovy:check docs:check`. Because all of those modules have reindex based tests, sadly.

nik9000 commented:
I left a few more comment, including a failing test. I'm pretty sure we're close to the end on this. I expect those are the last changes I'm likely to have any cause to suggest. After this is merged we'll need to do a few more things to finish the "delete-by-query porting project". This is my guess of the list:

* Remove the delete-by-query plugin. Should we just remove it in 5.0.0, call it a breaking change, and tell people to use the version in the reindex module? Is that ok because we're *kind of* just moving the API? It is losing some small functionality (timeouts) but it is gaining cancel, throttle, and rethrottle.
* Document all the breaking changes.
* Consolidate all those tests that have a version of reindex and a version for update by query into single tests.
** Add a delete-by-query case to the new retry tests. That should probably wait until #18456 is merged.

nik9000 commented:
Oh! More things:
* (optional) Implement `ctx.op = "delete"` on _update_by_query and _reindex.
* Cut all the script based cancel tests over to your listener based one. That is nicer.

tlrx commented:
@nik9000 Thanks again! I updated the code and also fixed the failing test.

Once this one is merged, I'll address the remaining points one after one :)

nik9000 commented:
> Once this one is merged, I'll address the remaining points one after one :)

I'm happy to grab a few!

nik9000 commented on docs/reference/docs/delete-by-query.asciidoc:
```diff
@@ -31,15 +31,18 @@ That will return something like this:
   "timed_out": false,
-  "deleted": 120,
-  "batches": 2,
+  "deleted": 119,
```
119 seems weird. Maybe we're missing a refresh?

nik9000 commented on docs/reference/docs/delete-by-query.asciidoc:
```diff
@@ -0,0 +1,318 @@
+{
+  "took" : 147,
+  "timed_out": false,
+  "deleted": 119,
```
119 feels wrong, like we're missing a refresh when we build the test data or something.

nik9000 replied:
Oh! I see why, because of that one other tweet. Got it.

tlrx replied:
Yes :)

nik9000 commented:
LGTM

natelapp commented:
Will this change also support deleting `external` versioned document that was broken by the move to the plugin?  The details of that bug are found in https://github.com/elastic/elasticsearch/issues/16654.

tlrx commented:
@natelapp In the current state, no. Your issue is a specific case and I created #18750 to handle it.
